ArjunaJTA/jta/classes/com/arjuna/ats/jta/common/JTAEnvironmentBean.java

```diff
@@ -1243,6 +1243,42 @@ public void setNotifyCommitMarkableResourceRecoveryModuleOfCompleteBranches(
 		this.notifyCommitMarkableResourceRecoveryModuleOfCompleteBranches = notifyCommitMarkableResourceRecoveryModuleOfCompleteBranches;
 	}
 
+    /**
+     * Remove a resource from the list of Markable Resources
+     */
+    public void removeCommitMarkableResource(String jndiName) {
+        synchronized (this) {
+            commitMarkableResourceJNDINames.remove(jndiName);
+            commitMarkableResourceTableNameMap.remove(jndiName);
+            performImmediateCleanupOfCommitMarkableResourceBranchesMap.remove(jndiName);
+            commitMarkableResourceRecordDeleteBatchSizeMap.remove(jndiName);
+        }
+    }
+
+    /**
+     * Add a resource to the list of Markable Resources
+     */
+    public void addCommitMarkableResource(String jndiName, String tableName, boolean immediateCleanup, int batchSize) {
+        synchronized (this) {
+            commitMarkableResourceJNDINames.add(jndiName);
+            commitMarkableResourceTableNameMap.put(jndiName, tableName);
+            performImmediateCleanupOfCommitMarkableResourceBranchesMap.put(jndiName, immediateCleanup);
+            commitMarkableResourceRecordDeleteBatchSizeMap.put(jndiName, batchSize);
+
+            // use the standard setter for notifyCommitMarkableResourceRecoveryModuleOfCompleteBranches,
+            // commitMarkableResourceRecordDeleteBatchSize and commitMarkableResourceTableName
+        }
+    }
+
+    public CmrJndiConfig getCmrJndiConfig() {
```

Comment:
I can't get the reason of having this new DTO `CmrJndiConfig`. Is that for more easily gathering the CMR data in the user code or is there a different reason?

Reply:
The properties in the DTO are all dependent on each other and need to be set as a group. Therefore they also need to be read as a group under the protection of the JTAEnvironmentBean lock. The methods `removeCommitMarkableResource`, `addCommitMarkableResource` and `getCmrJndiConfig` all need to be protected by this lock and this is achieved by copying them into the DTO.

BTW the DTO is still caching the collections. I have just pushed an update to the CmrJndiConfig constructor which avoids the caching. Good spot.

Reply:
I see, with the copy constructor this makes sense now.

But I don't like the approach in general. I could be missing something but let me phrase by objections.

* These new methods changes the way how the `JTAEnvironmentBean` has been used. There were no methods which groups some operations so far (if I don't count the tightly coupled attributes with `*ClassNames`). Bean used the setter and getters and that was all. The Narayana machinery of the injecting of properties expects to call the setters and getters.
* The patch seems to be tailored only for the current WFLY issue where we expect only 4 properties to be set up and which are needed to be guarded by lock. But `JTAEnvBean` may be used in different runtimes.
* There is specified no contract of the class how it should be used regards of these new methods. If there is a new integrator then how to get the right way work with CMR properties.
* Despite the lock guards the `add`, `remove` and `get` there is not thread safety ensured. The arrays and maps may be changed in the background by another thread. What thread safety is tried to be ensured by the lock? Should not be lock defined at the caller side for the block? (https://github.com/wildfly/wildfly/blob/master/transactions/src/main/java/org/jboss/as/txn/service/CMResourceService.java#L72-L77)
* If those CMR properties are dependent and they should be changed atomically would not be better to define some (immutable) DTO which gather those values and may be switched in some map/list atomically?

Reply:
Context: the JTA config bean maintains a list of CM resources, each identified by a unique JNDI name. Currently it does this via a number of collections.

I introduce new methods `add`, `remove` and `get` to manage these collections -  the other getters and setters in the config bean need to be deprecated, ie it is not tailored to the WFLY issue (this addresses your 1st, 2nd and 3rd points). Therefore the CMR propertes are managed as group and my PR does this (this addresses your 5th point).

When a new resource is added we need to update four maps atomically where each map is keyed by the unique resource name (this addresses your 4th point). When a reader gets the config the constructor for the DTO takes a copy of the four collections, the reader can then search the collections for the CM resource it is interested in (with no concern about thread safety).

The only point that I may not have addressed is your statement: `The Narayana machinery of the injecting of properties`. I don't see where that applies to the CMR functionality and why my PR breaks it.

Reply:
I think I see part of your concern in point 1 about the `Narayana injection machinery`, that relates to reading the properties from XML files. Let me check our XML parser and see if it can support this new technique.

Reply:
I understand.